receivers-in-go.md

```diff
@@ -0,0 +1,257 @@
+will be GREAT!.. or will it?
+
+
+## Facts about methods and receivers
+
+While we can call a method on a type instance and get the receiver implicitly, 
+it can also be called explicitly:
+
+```go
+myServer := &Server{}
+Server.Name(myServer)
+(*Server).Close(&myServer)
+```
+
+These functions can be passed around as references just like any other function, 
+with the implicit receiver or without:
+
+```go
+withReceiver := myServer.Name
+without := Server.Name
+```
+
+Receivers can be passed by reference or passed by value.
+
+```go
+func (byValue Server) Hello() { ... }
+func (byReference *Server) Bye() { ... }
+```
+
+This is to illustrate that struct methods in Go are merely thin sugar over 
+traditional C-style struct helper declarations, an equivalent C method might 
+look like this:
+
+```c
+void server_close(server *srv) { ... }
+```
+
+Go helps by namespacing the methods and implicitly passing the receiver when 
+called on an instance, but otherwise there is very little magic going on.
+
+In other languages where `this` and `self` is a thing (Python, Ruby, JavaScript, 
+and so on) it's a much more complicated situation. These are not vanilla local 
+variables wearing fancy pants. The thing we might expect `this` to represent 
+inside of a method could actually represent something very different once 
+inheretance or metaclasses had their way. In effect, it might not make any sense 
+to give contextual names to `self` in Python, but it definitely makes sense in 
+Go.
+
+## Reshaping our code
+
+As we start refactoring our code, we take a chunk of code that is already 
+functional and put it in another context where it allows for more flexibility 
+towards a higher-level goal.
+
+The first step of taking flat code and embedding it into a struct method is 
+usually fine regardless of what the receiver is named. At least in the 
+beginning.
+
+Later, once we start moving pieces of code *between* levels abstractions or from 
+higher levels of abstraction into a lower level, this is when consistently 
+well-named receivers make a huge difference.
+
+Imagine taking this snippet from a higher-level container like `Room` which 
+holds groups of users in a server, and moving it up or down one level:
+
+```go
+func (room *Room) Announce() {
+    srv := room.Server()
+    for _, c := range srv.Clients() {
+        // Send announcement to all clients about a new room
+        c.Send(srv.RenderAnnouncement(room))
+    }
+}
+
+// Moved between...
+
+func (srv *Server) AddRoom(room *Room) {
+    for _, c := range srv.Clients() {
+        // Send announcement to all clients about a new room
+        c.Send(srv.RenderAnnouncement(room))
+    }
+}
+```
+
+This is a great little pattern to keep everything working despite moving between 
+layers of abstraction. Note how the inner code stays identical and all we're 
+doing is adding a little extra context outside of it.
+
+On the other hand:
```

Comment:
I do think you should lead with this instead of starting with the right way. Problem, meet solution! Brings a little more drama. In the current order, we already feel like we know the answer, because we're smart hackers and we don't need no andrey to tell us what to do